VisualPinball.Unity/VisualPinball.Unity.Editor/Layers/LayerTreeElement.cs

```diff
@@ -0,0 +1,209 @@
+using System.Collections.Generic;
+using System.Linq;
+using UnityEditor;
+using UnityEngine;
+using VisualPinball.Engine.VPT.Table;
+using VisualPinball.Unity.Editor.Utils.TreeView;
+using VisualPinball.Unity.VPT;
+
+namespace VisualPinball.Unity.Editor.Layers
+{
+	/// <summary>
+	/// Enum for LayerTreeElement visibility status (including children)
+	/// </summary>
+	internal enum LayerTreeElementVisibility
+	{
+		Hidden,			// Element is hidden and children also
+		Hidden_Mixed,	// Element is hidden but some children are visible
+		Visible_Mixed,	// Element is visible but some children are hidden
+		Visible			// Element is visible and children also
+	}
+
+	/// <summary>
+	/// Element type used in TreeModel for the LayerTreeView
+	/// </summary>
+	/// <remarks>
+	/// Could be initialized in different ways to fit all possible LayerTreeView item types
+	/// </remarks>
+	internal class LayerTreeElement : TreeElement
+	{
+		/// <summary>
+		/// Name of the layer
+		/// </summary>
+		public string LayerName;
+		public ILayerableItemBehavior Item { get; }
+
+		private readonly Table _table;
+
+		public LayerTreeViewElementType Type =>
+			_table != null
+				? LayerTreeViewElementType.Table
+				: Item != null
+					? LayerTreeViewElementType.Item
+					: LayerName != null ? LayerTreeViewElementType.Layer : LayerTreeViewElementType.Root;
+
+		private bool _isVisible = true;
+		public bool IsVisible 
+		{
+			get {
+				if (Type == LayerTreeViewElementType.Item) {
+					if (Item is Behaviour behaviour) {
+						_isVisible = !SceneVisibilityManager.instance.IsHidden(behaviour.gameObject);
+					}
+					else {
+						_isVisible = false;
+					}
+				}
+				return _isVisible;
+			}
+
+			set {
+				_isVisible = value;
+				if (Type == LayerTreeViewElementType.Item) {
+					var behaviour = Item as MonoBehaviour;
+					if (behaviour != null) {
+						Undo.RecordObject(behaviour, $"{behaviour.name} : Switch visibility to {_isVisible}.");
+					}
+					Item.EditorLayerVisibility = _isVisible;
+					if (behaviour != null) {
+						if (_isVisible) {
+							SceneVisibilityManager.instance.Show(behaviour.gameObject, true);
+						}
+						else {
+							SceneVisibilityManager.instance.Hide(behaviour.gameObject, true);
+						}
+					}
+				}
+				else if (Type != LayerTreeViewElementType.Root) {
+					if (HasChildren) {
+						foreach (LayerTreeElement child in Children) {
+							child.IsVisible = _isVisible;
+						}
+					}
+				}
+			}
+		}
+
+		public override string Name
+		{
+			get {
+				if (_table != null) {
+					return _table.Name;
+				}
+
+				if (Item is IIdentifiableItemBehavior identifiable) {
+					return identifiable.Name;
+				}
+
+				if (LayerName != null) {
+					return LayerName;
+				}
+
+				return "<Root>";

```

Comment:
Can't we rely on the `Type` here as well (like below)?

Reply:
Yup sure you're right !

Reply:
Cool, I've pushed a few changes meanwhile. And I'm nearly through!